modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RocksDbKeyValueStorage.java

```diff
@@ -1664,4 +1664,17 @@ public void unregisterRevisionUpdateListener(RevisionUpdateListener listener) {
     public CompletableFuture<Void> notifyRevisionUpdateListenerOnStart(long newRevision) {
         return watchProcessor.notifyUpdateRevisionListeners(newRevision);
     }
+
+    @Override
+    public void advanceSafeTime(HybridTimestamp newSafeTime) {
+        rwLock.writeLock().lock();
```

Comment:
Why do we need to take the lock here?

Reply:
`WatchProcessor.advanceSafeTime()` chains `notificationFuture` with its previous version, this must be done under a lock. `WatchProcessor.notifyWatches()` does the same, and its correctness is backed by taking the write lock